### PR TITLE
Feature 642 distance map tb

### LINKED ISSUE #642: Add Distance Map use-case for Brightness Temperature
*Replace italics below with details for this issue.*

## Describe the New Feature ##
Part of the UW-CIMSS project includes developing a use-case for the evaluation brightness temperature to compute Distance Map metrics.  Sarah Griffin has provided sample data and the Grid-Stat configuration she used for the initial analysis.  They can be found on kiowa:/d1/personal/jensen/JTTI-UW

### Acceptance Testing ###
OU-CAPS ensemble members and Satellite brightness temperature originally from:
Now found at kiowa:/d1/personal/jensen/JTTI_UW
New functionality:  Need to sort out how to read in UW_CIMSS netCDF format.  

Running MET independently using the original Grid_Stat config used by UW_CIMSS and commands generated by METplus works.
Running MET through METplus using the original Grid_Stat config used by UW_CIMSS works
Running MET through METplus using a slightly modified Grid_Stat config with 1 env variable passed from METplus .conf to MET config does not work.

### Time Estimate ###
1 day to get working
1-2 day to add use-case

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [x ] Get Filename template working
- [ ] Wrap MET config file
- [ ] Define sample data tarball
- [ ] Add use-case to CAM application directory 

### Relevant Deadlines ###
November 2020

### Funding Source ###
2770061

### Assignee ###
- [x ] Select **engineer(s)** or **no engineer** required
- [ x ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [ ] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by CPKalb**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
       The cases were run and the output was verified to be identical to the original script

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
       Run the case to be sure it works and compare netcdf distance maps output to original

- [x] Do these changes include sufficient documentation and testing updates? **[Yes or No]**
       Yes

- [x] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
    Yes, there will be new output data for the new cases

## Pull Request Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, and **Milestone**
- [x] After submitting the PR, select **Linked Issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: parm/met_config/MODEConfig_wrapped [hunk 1]
```diff
@@ -237,7 +237,7 @@ nc_pairs_flag   = {
    object_raw   = TRUE;
    object_id    = TRUE;
    cluster_id   = TRUE;
-   polylines    = TRUE;
+   polylines    = FALSE;
```
Comment (georgemccabe) [inline parm/met_config/MODEConfig_wrapped]: Why was this changed? This change will likely break all of the existing use cases that run MODE because the output will differ. If you need to override this value for a specific use case, you can set this value in the .conf file:

MODE_MET_CONFIG_OVERRIDES = nc_pairs_flag = {polylines    = FALSE;}

File: parm/use_cases/model_applications/convection_allowing_models/GridStat_fcstFV3_obsGOES_BrightnessTempDmap.conf [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+FCST_IS_PROB = false
+
+# Grid stat variables
+FCST_GRID_STAT_VAR1_NAME = SBTA1613_topofatmosphere
+FCST_GRID_STAT_VAR1_LEVELS = "(*,*)"
+FCST_GRID_STAT_VAR1_THRESH = le235
+FCST_GRID_STAT_VAR1_OPTIONS = file_type = NETCDF_MET;
+
+FCST_IS_PROB = false
```
Comment (georgemccabe) [inline parm/use_cases/model_applications/convection_allowing_models/GridStat_fcstFV3_obsGOES_BrightnessTempDmap.conf]: This line is a duplicate of line 23. Please remove one of them.

File: parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+MODE_CONV_RADIUS = 5
+
+MODE_CONV_THRESH = <=235
+
+MODE_MERGE_THRESH = <=235
+
+MODE_MERGE_FLAG = NONE
+
+MODE_MASK_POLY = {PARM_BASE}/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs/CentUS.nc
```
Comment (georgemccabe) [inline parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf]: This file should be added to the sample input data and referenced from there instead of being put in the repository. It is too big!

**Review by georgemccabe — CHANGES_REQUESTED** [review]
There are a few things that should be updated. See the notes on the files.

File: parm/met_config/MODEConfig_wrapped (hunk 1 above)
Comment (CPKalb) [reply parm/met_config/MODEConfig_wrapped]: This was changed to make the files CF compliant so they can be easily read back into MET.   Let me try doing this, as I've never done this.  Do I need to make edits to the config file for this to work?

File: parm/use_cases/model_applications/convection_allowing_models/GridStat_fcstFV3_obsGOES_BrightnessTempDmap.conf (hunk 2 above)
Comment (CPKalb) [reply parm/use_cases/model_applications/convection_allowing_models/GridStat_fcstFV3_obsGOES_BrightnessTempDmap.conf]: Will do

File: parm/met_config/MODEConfig_wrapped (hunk 1 above)
Comment (georgemccabe) [reply parm/met_config/MODEConfig_wrapped]: Actually it looks like your use case doesn't even use this file, so you should be able to just change this value back to the default.

Comment (CPKalb) [reply parm/met_config/MODEConfig_wrapped]: The use case does use the output netcdf file

Comment (georgemccabe) [reply parm/met_config/MODEConfig_wrapped]: The use case uses this file:

MODE_CONFIG_FILE = {CONFIG_DIR}/MODEConfig_BrightnessTemp

not parm/met_config/MODEConfig_wrapped

File: parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf (hunk 3 above)
Comment (CPKalb) [reply parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf]: How do I add a mask file to the sample data?  The same way that I would add a data tarball?

File: parm/met_config/MODEConfig_wrapped (hunk 1 above)
Comment (CPKalb) [reply parm/met_config/MODEConfig_wrapped]: Hmmm, I thought I only changed this in the case use file, not MODEConfig_wrapped.  LEt me check on this

File: parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf (hunk 3 above)
Comment (georgemccabe) [reply parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf]: You add it to the directory on mohawk. There are steps to follow in the Contributor's Guide. I can take care of this for you now. The disk on mohawk is at 100% so some of the instructions won't work as expected.

File: parm/met_config/MODEConfig_wrapped (hunk 1 above)
Comment (CPKalb) [reply parm/met_config/MODEConfig_wrapped]: Yeah, that was an error then.  I'm changing it back.  I'm testing my changes and then will push a new commit

File: parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf (hunk 3 above)
Comment (CPKalb) [reply parm/use_cases/model_applications/convection_allowing_models/MODE_fcstFV3_obsGOES_BrightnessTempObjs.conf]: Okay.  Thanks

**Comment by georgemccabe** [comment]
@CPKalb could you please fill out the template for reviewing this pull request?

**Comment by CPKalb** [comment]
What is the test suite?

**Comment by georgemccabe** [comment]
> 
> 
> What is the test suite?

The automated tests run in GitHub Actions. This PR will change the output because there will be new output data from the new cases.

**Comment by georgemccabe** [comment]
An update on this review: I looked over the changes and they look good. Both cases ran properly on kiowa. The use cases have their own MET config file, which we are moving away from in favor of using GRID_STAT_MET_CONFIG_OVERRIDES and MODE_MET_CONFIG_OVERRIDES. I made some changes to do this, however there are changed in #768 that will be needed to switch to the new format for MODE because grid_res is set in this case and that needs to be set via a METplus config. Once the changes for #768 are merged into develop, I can test that the MODE case runs as expected and approve this PR.

**Comment by CPKalb** [comment]
Okay.  I believe that grid_res was set, as well as some of the mode specific parameters in the function to identify objects.  Let me know if there is anything I can do to help.

**Review by georgemccabe — CHANGES_REQUESTED** [review]
I made the updates to the config files but 1 of the cases failed due to a typo. It is re-running now. If all of the jobs succeed in the push run for this branch, then I will approve this PR. The pull request run will have differences because there is new data and the names of one of the use case groups changed (from 6+ to 6).

Push run:
https://github.com/dtcenter/METplus/actions/runs/641300331

Pull request run:
https://github.com/dtcenter/METplus/actions/runs/641300429

**Comment by georgemccabe** [comment]
Update: I confirmed that the new use cases run successfully. I am updating the logic for running the use case tests so that it will be easier to see that the only change to the test output is the addition of new data from this use case. I will apply those changes to this branch and approve.

**Review by georgemccabe — APPROVED** [review]
Approving this PR and creating a new one from branch feature_642_test_ci_updates for the changes to GitHub Actions
